From a42346b599f5ebba2d0ac7bb37cda7f1f242f641 Mon Sep 17 00:00:00 2001
From: Dung Cao <dung@os.amperecomputing.com>
Date: Tue, 20 Aug 2024 16:38:36 +0700
Subject: [PATCH] Support restore FaultLog entries after service restart

Implemented serialization of faultlog entry attributes using the
nlohmann::json library. Serializing attributes including type,
primaryLogId, and additionalTypeName.

Serialize and persist the FaultLog Dbus object when creating a new
entry. De-serialize and restore them (if persistent ones exist) when
the phosphor-debug-collector starts up.

Upstream-Status: Pending

Signed-off-by: Dung Cao <dung@os.amperecomputing.com>
---
 dump_entry.cpp            |   2 +
 dump_manager_faultlog.cpp | 171 ++++++++++++++++++++++++++++++--------
 dump_manager_faultlog.hpp |  12 +--
 faultlog_dump_entry.cpp   | 121 +++++++++++++++++++++++++++
 faultlog_dump_entry.hpp   |  25 ++++++
 5 files changed, 292 insertions(+), 39 deletions(-)

diff --git a/dump_entry.cpp b/dump_entry.cpp
index b7fa3a0..027c1db 100644
--- a/dump_entry.cpp
+++ b/dump_entry.cpp
@@ -91,6 +91,7 @@ void Entry::serialize()
         j["originatorId"] = originatorId();
         j["originatorType"] = originatorType();
         j["startTime"] = startTime();
+        j["elapsed"] = elapsed();
 
         os << j.dump();
     }
@@ -143,6 +144,7 @@ void Entry::deserialize(const std::filesystem::path& dumpPath)
                 originatorId(j["originatorId"].get<std::string>());
                 originatorType(j["originatorType"].get<originatorTypes>());
                 startTime(j["startTime"].get<uint64_t>());
+                elapsed(j["elapsed"].get<uint64_t>());
             }
             else
             {
diff --git a/dump_manager_faultlog.cpp b/dump_manager_faultlog.cpp
index 7d03c8b..5c96e56 100644
--- a/dump_manager_faultlog.cpp
+++ b/dump_manager_faultlog.cpp
@@ -87,46 +87,30 @@ sdbusplus::message::object_path Manager::createDump(
         "next entry id: {ID}, entries.size(): {SIZE}",
         "ID", id, "SIZE", entries.size());
 
-    std::filesystem::path faultLogFilePath(
-        std::string(FAULTLOG_DUMP_PATH) + idString);
-    std::ofstream faultLogFile;
-
-    errno = 0;
-
-    faultLogFile.open(faultLogFilePath,
-                      std::ofstream::out | std::fstream::trunc);
-
-    if (faultLogFile.is_open())
+    std::filesystem::path faultLogFilePath = primaryLogIdStr;
+    uint64_t fileSize = 0;
+    if (std::filesystem::exists(faultLogFilePath))
     {
-        lg2::info("faultLogFile is open");
-
-        faultLogFile << "This is faultlog file #" << idString << " at "
-                     << std::string(FAULTLOG_DUMP_PATH) + idString << std::endl;
-
-        faultLogFile.close();
+        fileSize = std::filesystem::file_size(faultLogFilePath);
     }
-    else
+    lg2::info("file_size: {SIZE}", "SIZE", fileSize);
+    std::filesystem::path filePath = dumpDir + idString;
+    if (!std::filesystem::exists(filePath))
     {
-        lg2::error(
-            "Failed to open fault log file at {FILE_PATH}, errno: {ERRNO}, "
-            "strerror: {STRERROR}, OBJECTPATH: {OBJECT_PATH}, ID: {ID}",
-            "FILE_PATH", faultLogFilePath, "ERRNO", errno, "STRERROR",
-            strerror(errno), "OBJECT_PATH", objPath, "ID", id);
-        elog<Open>(ErrnoOpen(errno), PathOpen(objPath.c_str()));
+        std::filesystem::create_directory(filePath);
     }
-
+    filePath /= FAULTLOG_FILE;
     try
     {
         lg2::info("dump_manager_faultlog.cpp: add faultlog entry");
-
-        entries.insert(std::make_pair(
-            id,
-            std::make_unique<faultlog::Entry>(
-                bus, objPath.c_str(), id, generateTimestamp(),
-                std::filesystem::file_size(faultLogFilePath), faultLogFilePath,
-                phosphor::dump::OperationStatus::Completed,
-                originatorId, originatorType, entryType,
-                primaryLogIdStr, additionalTypeStr, *this, &entries)));
+        auto e = std::make_unique<faultlog::Entry>(
+                      bus, objPath.c_str(), id, generateTimestamp(),
+                      fileSize, filePath,
+                      phosphor::dump::OperationStatus::Completed,
+                      originatorId, originatorType, entryType,
+                      primaryLogIdStr, additionalTypeStr, *this, &entries);
+        e->serializeEntry();
+        entries.insert(std::make_pair(id, std::move(e)));
     }
     catch (const std::invalid_argument& e)
     {
@@ -155,6 +139,15 @@ void Manager::deleteAll()
 
     removeAllDataEntry();
 
+    // Delete the persistent representation of all FaultLog entries.
+    for ( auto it = entries.begin(); it != entries.end(); ++it  )
+    {
+        fs::path faultlogPath(FAULTLOG_DUMP_PATH);
+        uint32_t id = it->first;
+        faultlogPath /= std::to_string(id);
+        fs::remove(faultlogPath);
+    }
+
     phosphor::dump::Manager::deleteAll();
 
     auto iter = savedCperLogEntries.begin();
@@ -631,9 +624,121 @@ void Manager::removeEarliestEntry(std::string &additionalTypeStr)
         }
     }
 
+    /* Delete the persistent representation of this FaultLog entry */
+    fs::path faultlogPath(FAULTLOG_DUMP_PATH);
+    uint32_t id = it->first;
+    faultlogPath /= std::to_string(id);
+    fs::remove(faultlogPath);
+
+    /* Delete the FaultLog entry */
     entries.erase(it);
 }
 
+void Manager::restore()
+{
+    lg2::info("dump_manager_faultlog restore is called");
+    fs::path dir(FAULTLOG_DUMP_PATH);
+    if (!fs::exists(dir) || fs::is_empty(dir))
+    {
+        return;
+    }
+
+    for (auto& file : fs::directory_iterator(dir))
+    {
+        auto id = file.path().filename().c_str();
+        auto idNum = std::stol(id);
+        auto idString = std::to_string(idNum);
+        auto objPath = std::filesystem::path(baseEntryPath) / idString;
+        std::filesystem::path filePath = file.path();
+        FaultDataType entryType;
+        std::string primaryLogIdStr;
+        std::string additionalTypeStr;
+        std::string originatorId;
+        originatorTypes originatorType;
+
+        filePath /= FAULTLOG_FILE;
+        auto e = std::make_unique<faultlog::Entry>(
+                      bus, objPath.c_str(), idNum, generateTimestamp(),
+                      0, filePath,
+                      phosphor::dump::OperationStatus::Completed,
+                      originatorId, originatorType, entryType,
+                      primaryLogIdStr, additionalTypeStr, *this, &entries);
+
+        e->deserializeEntry();
+        e->serializeEntry();
+        entries.insert(std::make_pair(idNum, std::move(e)));
+    }
+
+    if (!entries.empty())
+    {
+        // Restore the counter of last entry, cper and crashdump
+        restoreCounter();
+    }
+}
+
+void Manager::restoreCounter()
+{
+    lastEntryId = entries.rbegin()->first;
+
+    for ( auto it = entries.begin(); it != entries.end(); ++it  )
+    {
+        faultLogSize++;
+        if (faultLogSize > MAX_TOTAL_FAULT_LOG_ENTRIES)
+        {
+            faultLogSize = MAX_TOTAL_FAULT_LOG_ENTRIES;
+        }
+        auto secondPtr = it->second.get();
+        FaultDataType entryType =
+                dynamic_cast<faultlog::Entry*>(secondPtr)->type();
+        std::string primaryLogId =
+                dynamic_cast<faultlog::Entry*>(secondPtr)->primaryLogId();
+        std::string additionalTypeStr =
+                dynamic_cast<faultlog::Entry*>(secondPtr)->additionalTypeName();
+        // Remove fault log file
+        switch (entryType) {
+        case FaultDataType::CPER:
+            cperLogSize++;
+            if (cperLogSize > MAX_TOTAL_CPER_LOG_ENTRIES)
+            {
+                cperLogSize = MAX_TOTAL_CPER_LOG_ENTRIES;
+            }
+            break;
+        case FaultDataType::Crashdump:
+            crashdumpSize++;
+            if (crashdumpSize > MAX_TOTAL_CRASHDUMP_ENTRIES)
+            {
+                crashdumpSize = MAX_TOTAL_CRASHDUMP_ENTRIES;
+            }
+            if (!additionalTypeStr.empty())
+            {
+                /* OEM */
+                if (additionalTypeStr == "BERT")
+                {
+                    bertSize++;
+                    if (bertSize > MAX_TOTAL_BERT_ENTRIES)
+                    {
+                        bertSize = MAX_TOTAL_BERT_ENTRIES;
+                    }
+                }
+                if (additionalTypeStr == "Diagnostic")
+                {
+                    diagnosticSize++;
+                    if (diagnosticSize > MAX_TOTAL_DIAGNOSTIC_ENTRIES)
+                    {
+                        diagnosticSize = MAX_TOTAL_DIAGNOSTIC_ENTRIES;
+                    }
+                }
+            }
+            break;
+        default:
+            lg2::error("Incorrect FaultLog Entry Type");
+            elog<InternalFailure>();
+            break;
+        }
+    }
+}
+
+
 } // namespace faultlog
 } // namespace dump
 } // namespace phosphor
diff --git a/dump_manager_faultlog.hpp b/dump_manager_faultlog.hpp
index 94cb87d..e04afdd 100644
--- a/dump_manager_faultlog.hpp
+++ b/dump_manager_faultlog.hpp
@@ -11,6 +11,7 @@
 #include <sdbusplus/bus.hpp>
 #include <sdbusplus/bus/match.hpp>
 #include <sdbusplus/server/object.hpp>
+#include <filesystem>
 #include <xyz/openbmc_project/Dump/Create/server.hpp>
 
 namespace phosphor
@@ -22,6 +23,7 @@ namespace faultlog
 
 using CreateIface = sdbusplus::server::object_t<
     sdbusplus::xyz::openbmc_project::Dump::server::Create>;
+namespace fs = std::filesystem;
 
 /** @class Manager
  *  @brief FaultLog Dump manager implementation.
@@ -71,12 +73,7 @@ class Manager :
         diagnosticSize = 0;
     }
 
-    void restore() override
-    {
-        // TODO phosphor-debug-collector/issues/21: Restore fault log entries
-        // after service restart
-        lg2::info("dump_manager_faultlog restore not implemented");
-    }
+    void restore() override;
 
     /** @brief  Delete all fault log entries and their corresponding fault log
      * dump files */
@@ -168,6 +165,9 @@ class Manager :
     /** @brief Check threshold condition of FaultLog */
     void checkThresholdFaultLog(FaultDataType type,
                                 std::string &additionalTypeStr);
+
+    /** @brief Restore the counter of last entry, cper and crashdump */
+    void restoreCounter();
 };
 
 } // namespace faultlog
diff --git a/faultlog_dump_entry.cpp b/faultlog_dump_entry.cpp
index 9ff1eb3..31f4cbe 100644
--- a/faultlog_dump_entry.cpp
+++ b/faultlog_dump_entry.cpp
@@ -1,6 +1,7 @@
 #include "faultlog_dump_entry.hpp"
 
 #include <phosphor-logging/lg2.hpp>
+#include <nlohmann/json.hpp>
 
 namespace phosphor
 {
@@ -28,6 +29,126 @@ void Entry::delete_()
     parentMap->erase(id);
 }
 
+void Entry::serializeFaultLogEntry()
+{
+    // Folder for serialized entry
+    std::filesystem::path dir = file.parent_path() / PRESERVE;
+
+    // Serialized entry file
+    std::filesystem::path serializePath = dir / FAULTLOG_SERIALIZE_FILE;
+    try
+    {
+        if (!std::filesystem::exists(dir))
+        {
+            std::filesystem::create_directories(dir);
+        }
+
+        std::ofstream os(serializePath, std::ios::binary);
+        if (!os.is_open())
+        {
+            lg2::error("Failed to open file for serialization: {PATH} ", "PATH",
+                       serializePath);
+            return;
+        }
+        nlohmann::json j;
+        j["dumpId"] = id;
+        j["type"] = type();
+        j["primaryLogId"] = primaryLogId();
+        j["additionalTypeName"] = additionalTypeName();
+
+        os << j.dump();
+    }
+    catch (const std::exception& e)
+    {
+        lg2::error("Serialization error: {PATH} {ERROR} ", "PATH",
+                   serializePath, "ERROR", e);
+
+        // Remove the serialization folder if that got created
+        // Ignore the error since folder may not be created
+        std::error_code ec;
+        std::filesystem::remove_all(dir, ec);
+    }
+}
+
+void Entry::serializeEntry()
+{
+    // Serialize faultlog entry
+    serializeFaultLogEntry();
+    // Serialize base dump entry
+    serialize();
+}
+
+void Entry::derializeFaultLogEntry()
+{
+    try
+    {
+        // .preserve folder
+        std::filesystem::path dir = file.parent_path() / PRESERVE;
+        if (!std::filesystem::exists(dir))
+        {
+            lg2::info("Serialization directory: {SERIAL_DIR} doesnt exist, "
+                      "skip deserialization",
+                      "SERIAL_DIR", dir);
+            return;
+        }
+
+        // Serialized entry
+        std::filesystem::path serializePath = dir / FAULTLOG_SERIALIZE_FILE;
+        std::ifstream is(serializePath, std::ios::binary);
+        if (!is.is_open())
+        {
+            lg2::error("Failed to open file for deserialization: {PATH}",
+                       "PATH", serializePath);
+            return;
+        }
+        nlohmann::json j;
+        is >> j;
+
+        uint32_t storedId;
+        j.at("dumpId").get_to(storedId);
+        if (storedId == id)
+        {
+            type(j["type"].get<FaultDataType>());
+            primaryLogId(j["primaryLogId"].get<std::string>());
+            additionalTypeName(j["additionalTypeName"].get<std::string>());
+            uint64_t fileSize = 0;
+            std::filesystem::path faultLogFilePath;
+            faultLogFilePath = primaryLogId();
+            if (std::filesystem::exists(faultLogFilePath))
+            {
+                fileSize = std::filesystem::file_size(faultLogFilePath);
+            }
+            size(fileSize);
+        }
+        else
+        {
+            lg2::error("The id ({ID_IN_FILE}) is not matching the dump id "
+                       "({DUMPID}); skipping deserialization.",
+                       "ID_IN_FILE", storedId, "DUMPID", id);
+
+            // Id is not matching, this could be due to file corruption
+            // deleting the .preserve folder.
+            // Attempt to delete the folder and ignore any error.
+            std::error_code ec;
+            std::filesystem::remove_all(dir, ec);
+        }
+    }
+    catch (const std::exception& e)
+    {
+        lg2::error("Deserialization error: {PATH}, {ERROR}", "PATH",
+                   file.parent_path(), "ERROR", e);
+    }
+}
+
+void Entry::deserializeEntry()
+{
+    // Deserialize faultlog entry
+    derializeFaultLogEntry();
+    // Deserialize base dump entry
+    deserialize(file.parent_path());
+}
+
+
 } // namespace faultlog
 } // namespace dump
 } // namespace phosphor
diff --git a/faultlog_dump_entry.hpp b/faultlog_dump_entry.hpp
index 8fa0bad..c462b3a 100644
--- a/faultlog_dump_entry.hpp
+++ b/faultlog_dump_entry.hpp
@@ -27,6 +27,11 @@ using EntryIfaces = sdbusplus::server::object_t<
 using FaultDataType = sdbusplus::xyz::openbmc_project::Common::server::
     FaultLogType::FaultLogTypes;
 
+// Dummy file store the contents
+constexpr auto FAULTLOG_FILE = "dummy_file";
+
+// Serialize file store the contents of faultlog entry
+constexpr auto FAULTLOG_SERIALIZE_FILE = "faultlog_serialized_entry.json";
 
 /** @class Entry
  *  @brief OpenBMC Fault Log Dump Entry implementation.
@@ -80,10 +85,30 @@ class Entry : virtual public EntryIfaces, virtual public phosphor::dump::Entry
      */
     void delete_() override;
 
+    /** @brief Serialize the faultlog and base dump entry attributes
+     *  @return
+     */
+    void serializeEntry();
+
+    /** @brief Deserialize the faultlog and base dump entry attributes
+     *  @return
+     */
+    void deserializeEntry();
+
   private:
     /** @brief Fault log map containing this entry (e.g. the main fault log map
      * or a saved entries map) */
     std::map<uint32_t, std::unique_ptr<phosphor::dump::Entry>>* parentMap;
+
+    /** @brief Serialize the faultlog entry attributes
+     *  @return
+     */
+    void serializeFaultLogEntry();
+
+    /** @brief Deserialize the faultlog entry attributes
+     *  @return
+     */
+    void derializeFaultLogEntry();
 };
 
 } // namespace faultlog
-- 
2.34.1

